cmd/fips/main.go

```diff
@@ -0,0 +1,201 @@
+			packages.NeedName | packages.NeedFiles | packages.NeedCompiledGoFiles |
+			packages.NeedTypes | packages.NeedTypesInfo | packages.NeedTypesSizes,
+	}
+	pkgs, err := packages.Load(cfg, flag.Args()...)
+	if err != nil {
+		return nil, err
+	}
+	return pkgs, nil
+
+}
+
+func resolveFun(expr *ast.CallExpr, files []*ast.File) *ast.FuncDecl {
+	if ident, ok := expr.Fun.(*ast.Ident); ok {
+		if ident.Obj != nil {
+			fn, _ := ident.Obj.Decl.(*ast.FuncDecl)
+			return fn
+		}
+		// Ideally objects are always resolved by the package loader,
+		// but that's not always true so we try to resolve the function
+		// looking at other files within the same package.
+		for _, f := range files {
+			for _, d := range f.Decls {
+				if fn, ok := d.(*ast.FuncDecl); ok {
+					if fn.Name.Name == ident.Name {
+						return fn
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// hasBoringEnabled traverses the node to find if it has a boring.Enabled() call.
+// Function calls which are not yet resolved yet will be in files.
+func hasBoringEnabled(node ast.Node, seen map[*ast.FuncDecl]struct{}, files []*ast.File) bool {
+	var ret bool
+	ast.Inspect(node, func(n ast.Node) bool {
+		switch n := n.(type) {
+		case *ast.CallExpr: // enableBoring()
+			fn := resolveFun(n, files)
+			if fn != nil {
+				if _, ok := seen[fn]; ok {
+					return false
+				}
+				seen[fn] = struct{}{}
+				if hasBoringEnabled(fn, seen, files) {
+					ret = true
+					return false
+				}
+			}
+		case *ast.SelectorExpr: // foo.Fn(...)
+			if ident, ok := n.X.(*ast.Ident); ok {
+				if ident.Name == "boring" {
+					if n.Sel.Name == "Enabled" {
+						ret = true
+						return false
+					}
+				}
+			}
+		}
+		return true
+	})
+	return ret
+}
+
+// searchBoringCalls traverses the node searching for calls to the boring package.
+// Function calls which are not yet resolved will be searched in files.
+func searchBoringCalls(node ast.Node, files []*ast.File, calls []string) []string {
+	ast.Inspect(node, func(n ast.Node) bool {
+		switch n := n.(type) {
+		case *ast.CallExpr: // enableBoring()
+			fn := resolveFun(n, files)
+			if fn != nil {
+				calls = searchBoringCalls(fn.Body, files, calls)
+				return false
+			}
+		case *ast.SelectorExpr: // boring.Fn(...)
+			if ident, ok := n.X.(*ast.Ident); ok {
+				if ident.Name == "boring" && n.Sel.Name != "Enabled" {
+					calls = append(calls, n.Sel.Name)
+					return false
+				}
+			}
+		}
+		return true
+	})
+	return calls
+}
+
+func processFuncDecl(decl *ast.FuncDecl, files []*ast.File, report *fnReport) bool {
+	if decl.Body == nil {
+		return false
+	}
+	if decl.Recv != nil {
+		// TODO: check methods.
+		return false
+	}
+	seen := make(map[*ast.FuncDecl]struct{})
+	ast.Inspect(decl.Body, func(n ast.Node) bool {
+		switch stmt := n.(type) {
+		// Only search for boring calls that are inside a boring.Enabled if block.
+		case *ast.IfStmt:
+			if hasBoringEnabled(stmt.Cond, seen, files) { // if boring.Enabled() {...}
```

Comment:
I've simplified the `hasBoringEnabled` signature by removing `files`. Now the object resolution takes place before processing an exported function.

I cannot define it as a local function because it is recursive and the go compiler doesn't like it. I could try to refactor the whole algorithm to use methods instead of functions and then add the `seen` map to the parent `struct`, but IMO this will  add more complexity than just passing `seen` as parameter. Thoughts?

I'm not worried about the testability of internal methods, I'm using a black-box testing approach here.

Reply:
A local function can be recursive, it just needs to have the func variable declared first, to capture:

https://play.golang.org/p/OxHCMxc_S_G
```go
var x func(i int)
x = func(i int) {
	if i > 0 {
		fmt.Printf("%v\n", i)
		x(i - 1)
	}
}
```

But, simplifying works too. 😄

Reply:
FWIW, though:

> I could try to refactor the whole algorithm to use methods instead of functions [...]

I agree this would be more complexity than it's worth in this case.

Re-reading this bit of the code, now that there are only two args and `seen` is only used in two funcs anyway, I'm personally ok leaving it as you have it now.